src/com/amazon/corretto/crypto/provider/AesGcmSpi.java

```diff
@@ -307,103 +314,115 @@ protected void engineInit(
       final AlgorithmParameterSpec algorithmParameterSpec,
       final SecureRandom secureRandom)
       throws InvalidKeyException, InvalidAlgorithmParameterException {
-    if (key == null) {
-      throw new InvalidKeyException("Key can't be null");
-    }
 
-    final GCMParameterSpec spec;
-    if (algorithmParameterSpec instanceof GCMParameterSpec) {
-      spec = (GCMParameterSpec) algorithmParameterSpec;
-    } else if (algorithmParameterSpec instanceof IvParameterSpec) {
-      spec =
-          new GCMParameterSpec(
-              DEFAULT_TAG_LENGTH, ((IvParameterSpec) algorithmParameterSpec).getIV());
-    } else {
-      throw new InvalidAlgorithmParameterException(
-          "I don't know how to handle a " + algorithmParameterSpec.getClass());
-    }
+    final int opMode = checkOperation(jceOpMode);
 
-    byte[] encodedKey = null;
-    if (jceKey != key) {
```

Comment:
This check only saves time if the Cipher object is reused AND the same AES key is use AND the key is the same Java object; in all other cases, keeping `jceKey` around not beneficial. From security perspective, `jceKey` is a cryptographic material and it's not a good practice to keep a reference to a secret object for longer than it is needed.

Reply:
NOT A CONTRIBUTION

---

Consider the common case of a single Cipher object being used to encrypt/decrypt a lot of data with the same key but different parameters (iv or gcm). In this case you can absolutely expect to see the exact same Java key object provided each time. What's more, I would expect that the majority of the time that a cipher object is reused, it is reused with the same Java key object.

You're already keeping the cryptographic materials around in `private byte[] key`, so minimizing it here doesn't help. What's more, as you're calling `getEncoded()` each time for the comparison, you're resulting in more copies of the cryptographic material spilling to memory than before this change. (In other words, your change will result in more copies of the key in memory rather than fewer.)

---

NOT A CONTRIBUTION

Reply:
Despite not having any impact on performance, I brought back the logic. I'm not sure about "common" this pattern is, since we have many internal customers that prefer not reuse a Cipher object so that they don't have to deal with thread-safety issues.